lib/fluent/plugin/buffer.rb

```diff
@@ -456,9 +447,21 @@ def enqueue_unstaged_chunk(chunk)
         end
       end
 
+      def update_timekeys
+        synchronize do
```

Comment:
The `synchronize` here is not some Ruby keyword as I imagined. Please take a look - maybe there is something I could do to reduce the lock(s) needed. The method does have to loop through everything in stage and queue, though the consistency of data it reads doesn't matter (e.g. it's okay for the same buffer to be found in `stage` and `queue`).